include/boost/math/special_functions/jacobi_theta.hpp

```diff
@@ -0,0 +1,616 @@
+    RealType eps = policies::get_epsilon<RealType, Policy>();
+    RealType result = RealType(0);
+
+    // n>=0 even
+    result -= _jacobi_theta_sum(tau, z + constants::half_pi<RealType>(), constants::two_pi<RealType>(), eps);
+    // n>0 odd
+    result += _jacobi_theta_sum(tau, z + constants::half_pi<RealType>() + constants::pi<RealType>(), constants::two_pi<RealType>(), eps);
+    // n<0 odd
+    result += _jacobi_theta_sum(tau, z - constants::half_pi<RealType>(), -constants::two_pi<RealType>(), eps);
+    // n<0 even
+    result -= _jacobi_theta_sum(tau, z - constants::half_pi<RealType>() - constants::pi<RealType>(), -constants::two_pi<RealType>(), eps);
+
+    return result * sqrt(tau);
+}
+
+template <class RealType, class Policy>
+inline RealType
+_IMAGINARY_jacobi_theta2tau(RealType z, RealType tau, const Policy& pol) {
+    BOOST_MATH_STD_USING
+    RealType eps = policies::get_epsilon<RealType, Policy>();
+    RealType result = RealType(0);
+
+    // n>=0
+    result += _jacobi_theta_sum(tau, z + constants::half_pi<RealType>(), constants::pi<RealType>(), eps);
+    // n<0
+    result += _jacobi_theta_sum(tau, z - constants::half_pi<RealType>(), -constants::pi<RealType>(), eps);
+
+    return result * sqrt(tau);
+}
+
+template <class RealType, class Policy>
+inline RealType
+_IMAGINARY_jacobi_theta3tau(RealType z, RealType tau, const Policy& pol) {
+    BOOST_MATH_STD_USING
+    RealType eps = policies::get_epsilon<RealType, Policy>();
+    RealType result = 0;
+
+    // n=0
+    result += exp(-z*z*tau/constants::pi<RealType>());
+    // n>0
+    result += _jacobi_theta_sum(tau, z + constants::pi<RealType>(), constants::pi<RealType>(), eps);
+    // n<0
+    result += _jacobi_theta_sum(tau, z - constants::pi<RealType>(), -constants::pi<RealType>(), eps);
+
+    return result * sqrt(tau);
+}
+
+template <class RealType, class Policy>
+inline RealType
+_IMAGINARY_jacobi_theta4tau(RealType z, RealType tau, const Policy& pol) {
+    BOOST_MATH_STD_USING
+    RealType eps = policies::get_epsilon<RealType, Policy>();
+    RealType result = 0;
+
+    // n = 0
+    result += exp(-z*z*tau/constants::pi<RealType>());
+
+    // n > 0 odd
+    result -= _jacobi_theta_sum(tau, z + constants::pi<RealType>(), constants::two_pi<RealType>(), eps);
+    // n < 0 odd
+    result -= _jacobi_theta_sum(tau, z - constants::pi<RealType>(), -constants::two_pi<RealType>(), eps);
+    // n > 0 even
+    result += _jacobi_theta_sum(tau, z + constants::two_pi<RealType>(), constants::two_pi<RealType>(), eps);
+    // n < 0 even
+    result += _jacobi_theta_sum(tau, z - constants::two_pi<RealType>(), -constants::two_pi<RealType>(), eps);
+
+    return result * sqrt(tau);
+}
+
+// Begin public API
+
+// First Jacobi theta function (Parameterized by tau - assumed imaginary)
+// = 2 * Σ (-1)^n * exp(iπτ*(n+1/2)^2) * sin((2n+1)z)
+template <class RealType, class Policy>
+inline RealType
+jacobi_theta1tau(RealType z, RealType tau, const Policy& pol)
```

Comment:
Point of pedantry: I'd prefer it if the actual entry points followed the logic in https://www.boost.org/doc/libs/1_73_0/libs/math/doc/html/math_toolkit/special_tut/special_tut_impl.html, what we gain is the ability to handle arguments of different types (including integers), plus expression templates when used with multiprecision types.

On a related note, I'd like to see the entry points added to this test: https://github.com/boostorg/math/blob/develop/test/compile_test/instantiate.hpp,   

Also note that each special function appears in multiple sections in that file - you will very likely get test failures when doing this, but it does catch a lot of issues - including now, not forwarding policies correctly ;)

Reply:
To be clear - should I be using `promote_args` and friends in the public API? (I was advised against this previously.)

I've added the entry points (privately) to instantiate.hpp - lots of failures - will keep me busy tomorrow.

Reply:
How's it look now?

Reply:
>To be clear - should I be using promote_args and friends in the public API? (I was advised against this previously.)

Yeah, my bad, you had a complex number interface before and promote_args doesn't work (easily) with that.  Now that it's all real-valued, might as well be consistent with everything else.

aside 1: promote_args is mis-named, should really be "make_common_floating_point_type" or something.
aside 2: I guess it would help if I updated it to function with complex number types as well!

>I've added the entry points (privately) to instantiate.hpp - lots of failures - will keep me busy tomorrow.

Likely most will disappear once the argument type handling is in place.

>How's it look now?

You're way ahead of me, will catch up with this later!